src/abstractdataframe/join.jl

```diff
@@ -81,16 +81,622 @@ end
 
 Base.length(x::RowIndexMap) = length(x.orig)
 
-# composes the joined data table using the maps between the left and right
-# table rows and the indices of rows in the result
-
 _rename_cols(old_names::AbstractVector{Symbol},
              renamecols::Union{Function, Symbol, AbstractString},
              exclude::AbstractVector{Symbol} = Symbol[]) =
     Symbol[n in exclude ? n :
            (renamecols isa Function ? Symbol(renamecols(string(n))) : Symbol(n, renamecols))
            for n in old_names]
 
+struct OnColRow{T}
+    row::Int
+    cols::T
+
+    OnColRow(row::Union{Signed,Unsigned},
+             cols::NTuple{N, AbstractVector})  where {N} =
+        new{typeof(cols)}(Int(row), cols)
+end
+
+struct OnCol{T,N} <: AbstractVector{OnColRow{T}}
+    len::Int
+    cols::T
+
+    function OnCol(cs::AbstractVector...)
+        @assert length(cs) > 1
+        len = length(cs[1])
+        @assert all(x -> firstindex(x) == 1, cs)
+        @assert all(x -> lastindex(x) == len, cs)
+        new{typeof(cs), length(cs)}(len, cs)
+    end
+end
+
+Base.IndexStyle(::Type{<:OnCol}) = Base.IndexLinear()
+
+@inline Base.size(oc::OnCol) = (oc.len,)
+
+@inline function Base.getindex(oc::OnCol, i::Int)
+    @boundscheck checkbounds(oc, i)
+    return OnColRow(i, oc.cols)
+end
+
+@inline function Base.hash(ocr1::OnColRow{<:NTuple{2, AbstractVector}}, h::UInt)
```

Comment:
Have you tried with a generated function for this and `isequal`? Maybe the compiler would even be able to generate efficient code automatically?

Reply:
I was thinking about it, but I have experience that getting `@generated` code to work correctly with proper hygiene etc. is hard. I assumed that having more than 3 columns is super rare. I have checked non-generated approach like:
```
function hash(ocr1)
    r1 = ocr1.row
    hash(map(x -> (@inbounds x[r1]), length(ocr1.cols)))
end
```
or (I here show it for fixed number of parameters but just to show a general idea what I tried)
```
function h(ocr1)
           r1 = ocr1.row
           c = ocr1.cols
           @inbounds hash((c[1][r1], c[2][r1], c[3][r1]))
       end
```
and it is two times slower unfortunately.

I know that this code duplication is nasty. If you can propose a correct implementations of `hash`, `isequal`, and `isless` using `@generated` then I think it would be nice to replace them. For now I will add a comment that replacing them with `@generated` isa to-do (as it is for sure possible).